accepted/00000-jsonb-pillar-db-storage.md

```diff
@@ -0,0 +1,225 @@
+
+This RFC proposes to instead of plain files, database based storage should be used for all pillars produced by Uyuni.
+
+This is expected to help with maintaining consistency, help with backup procedure and with interserver sync as well, potential performance improvement (TBD - filesystem access vs. database indexes).
+
+# Detailed design
+[design]: #detailed-design
+
+## Database storage
+
+In this RFC I propose to use native JSON type of PostgreSQL ([see](https://www.postgresql.org/docs/12/datatype-json.html)) to use as a storage of the pillar data, particularly JSONB type:
+
+```sql
+CREATE TABLE suseSaltPillars
+(
+    target VARCHAR(256) NOT NULL,
+    category VARCHAR,
+    pillar JSONB
+);
+  
+CREATE INDEX suse_salt_pillar_target_idx ON suseSaltPillars (target);
+```
+
+To see differences between TEXT, JSON and JSONB datatype I quote PostgreSQL documentation:
+
+> JSON data types are for storing JSON (JavaScript Object Notation) data, as specified in RFC 7159.Such data can also be stored as text, but the JSON data types have the advantage of enforcing that each stored value is valid according to the JSON rules.
+>
+> The json and jsonb data types accept almost identical sets of values as input. The major practical difference is one of efficiency. The json data type stores an exact copy of the input text, which processing functions must reparse on each execution; while jsonb data is stored in a decomposed binary format that makes it slightly slower to input due to added conversion overhead, but significantly faster to process, since no reparsing is needed. jsonb also supports indexing, which can be a significant advantage.
+
+`target` column is mainly to contain `minion_id` of salt client, however to support universal and system group pillars I opted to use VARCHAR with limit of 256 (as is the limit of minion id, group name is limited to 64 characters). For universal pillar I propose to use `*` character.
+
+Added `category` VARCHAR is there to provide easer manipulation and understanding of stored data. Currently all Uyuni pillar data are written using `com.suse.manager.webui.services.pillar.MinionPillarManager` (assuming [PR#3093](https://github.com/uyuni-project/uyuni/pull/3093) is merged) and each type of pillars have their own respective [generator class](https://github.com/nadvornik/uyuni/tree/custominfo/java/code/src/com/suse/manager/webui/services/pillar). To maintain some form of distinction I propose this `category` column to allow to have multiple pillar entries per salt client.
```

Comment:
You might want to also add an index for the category - not necessary to specify here in the RFC, just a reminder for implementation time :wink:

Reply:
Added to the proposal.